packages/components/src/custom-gradient-picker/index.tsx

```diff
@@ -100,30 +118,58 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
 			onChange={ handleOnChange }
 			options={ GRADIENT_OPTIONS }
 			size="__unstable-large"
-			value={ hasGradient && type }
+			// NTS: Small runtime change. Previously, if `type` was undefined (which I don't think should ever happen) this check would return `false`, which would be an invalid value.
+			value={ hasGradient ? type : undefined }
 		/>
 	);
 };
 
-export default function CustomGradientPicker( {
+/**
+ * CustomGradientPicker is a React component that renders a UI for specifying
+ * linear or radial gradients. Radial gradients are displayed in the picker as
+ * a slice of the gradient from the center to the outside.
+ *
+ * ```jsx
+ * import { CustomGradientPicker } from '@wordpress/components';
+ * import { useState } from '@wordpress/element';
+ *
+ * const MyCustomGradientPicker = () => {
+ *   const [ gradient, setGradient ] = useState();
+ *
+ *   return (
+ *     <CustomGradientPicker
+ *			value={ gradient }
+ *			onChange={ setGradient }
+ *     />
+ *   );
+ * };
+ * ```
+ */
+export function CustomGradientPicker( {
 	/** Start opting into the new margin-free styles that will become the default in a future version. */
 	__nextHasNoMargin = false,
 	value,
 	onChange,
-	__experimentalIsRenderedInSidebar,
-} ) {
-	const gradientAST = getGradientAstWithDefault( value );
+	__experimentalIsRenderedInSidebar = false,
+}: CustomGradientPickerProps ) {
+	const { gradientAST, gradientAstValue } =
+		getGradientAstWithDefault( value );
 	// On radial gradients the bar should display a linear gradient.
 	// On radial gradients the bar represents a slice of the gradient from the center until the outside.
 	// On liner gradients the bar represents the color stops from left to right independently of the angle.
 	const background = getLinearGradientRepresentation( gradientAST );
-	const hasGradient = gradientAST.value !== DEFAULT_GRADIENT;
+	const hasGradient = gradientAstValue !== DEFAULT_GRADIENT;
 	// Control points color option may be hex from presets, custom colors will be rgb.
 	// The position should always be a percentage.
-	const controlPoints = gradientAST.colorStops.map( ( colorStop ) => ( {
-		color: getStopCssColor( colorStop ),
-		position: parseInt( colorStop.length.value ),
-	} ) );
+	const controlPoints = gradientAST.colorStops.map( ( colorStop ) => {
+		// Although it's already been checked by `hasUnsupportedLength` in `getGradientAstWithDefault`,
+		// TypeScript doesn't know that `colorStop.length` is not undefined here.
+		//NTS: Is there may be a more elegant way to help TS see that for itself? 🤔
+		return {
+			color: getStopCssColor( colorStop ),
+			position: parseInt( colorStop.length!.value ),
```

Comment:
Is there a more elegant way to help TS see for itself that `colorStop.length` is no longer undefined, and avoid using that `!`? 🤔

Reply:
Good idea. I took a stab at the second option of tweaking the return value. It meant using an `Exclude` to remove `undefined` as an option, and then a couple of the `DistributiveOmits` to selectively replace parent props as we work back up the tree.

<details>

<summary>Here's a diff of the proposed changes:</summary>

```diff

diff --git a/packages/components/src/custom-gradient-picker/index.tsx b/packages/components/src/custom-gradient-picker/index.tsx
index ada7b7392c..fda0df040a 100644
--- a/packages/components/src/custom-gradient-picker/index.tsx
+++ b/packages/components/src/custom-gradient-picker/index.tsx
@@ -162,7 +162,7 @@ export function CustomGradientPicker( {
 		// TypeScript doesn't know that `colorStop.length` is not undefined here.
 		return {
 			color: getStopCssColor( colorStop ),
-			position: parseInt( colorStop.length!.value ),
+			position: parseInt( colorStop.length.value ),
 		};
 	} );
 
diff --git a/packages/components/src/custom-gradient-picker/types.ts b/packages/components/src/custom-gradient-picker/types.ts
index 68d026eb7f..cc1e9fc105 100644
--- a/packages/components/src/custom-gradient-picker/types.ts
+++ b/packages/components/src/custom-gradient-picker/types.ts
@@ -125,3 +125,24 @@ export type InsertPointProps = {
 	disableAlpha: boolean;
 	__experimentalIsRenderedInSidebar: boolean;
 };
+
+// The following types are used to define a modified version of the
+// `gradientParser` package's `GradientNode` type. In this version,
+// the `length` property of the `ColorStops` contained in the node cannot be
+// `undefined`. The `getGradientAstWithDefault` and `hasUnsupportedLength` utils
+// ensure this is the case, but TypeScript couldn't detect the narrowing without
+// a more explicit type.
+type DefinedColorStopLength = Exclude<
+	gradientParser.ColorStop[ 'length' ],
+	undefined
+>;
+
+type ColorStopWithDefinedLength = DistributiveOmit<
+	gradientParser.ColorStop,
+	'length'
+> & { length: DefinedColorStopLength };
+
+export type GradientNodeWithDefinedStopLengths = DistributiveOmit<
+	gradientParser.GradientNode,
+	'colorStops'
+> & { colorStops: ColorStopWithDefinedLength[] };
diff --git a/packages/components/src/custom-gradient-picker/utils.ts b/packages/components/src/custom-gradient-picker/utils.ts
index 5b21f94c3e..b355949dc6 100644
--- a/packages/components/src/custom-gradient-picker/utils.ts
+++ b/packages/components/src/custom-gradient-picker/utils.ts
@@ -14,7 +14,7 @@ import {
 	DIRECTIONAL_ORIENTATION_ANGLE_MAP,
 } from './constants';
 import { serializeGradient } from './serializer';
-import type { ControlPoint } from './types';
+import type { ControlPoint, GradientNodeWithDefinedStopLengths } from './types';
 
 extend( [ namesPlugin ] );
 
@@ -75,7 +75,10 @@ export function getGradientAstWithDefault( value?: string ) {
 		gradientAstValue = serializeGradient( gradientAST );
 	}
 
-	return { gradientAST, gradientAstValue };
+	return { gradientAST, gradientAstValue } as {
+		gradientAST: GradientNodeWithDefinedStopLengths;
+		gradientAstValue: string;
+	};
 }
 
 export function getGradientAstWithControlPoints(

```

</details>

If it helps, try messing with this mock `GradientNode`:

```
const test: GradientNodeWithDefinedStopLengths = {
	type: 'linear-gradient',
	colorStops: [
		{
			type: 'literal',
			value: 'cross your fingers...',
			length: { type: '%', value: 'here we go...' },
		},
	],
};
```

With the above diff applied, if you play around with the different `ColorStop` types (`literal`, `hex`, `rgb`, or `rgba`), TypeScript should still expect the right `value` type for that `ColorStop` type, but it will never accept an `undefined` length. Thoughts? 😄

Reply:
Although this exploration is interesting and works as expected, my inclination would be to either keep the original `!` syntax, or to add a `@ts-expect-error` comment with the reason why.

Reply:
Added a `@ts-expect-error` ✅